File: android/app/src/main/java/dk/betterw4/android/feature/schedule/ClassNextLesson.kt
package dk.betterw4.android.feature.schedule

import java.time.LocalDateTime

/**
 * The next (or this week's first) timetable block for a class, from a week
 * that is already in the cache. No network.
 */
data class ClassNextLesson(
    val start: LocalDateTime,
    val room: String? = null,
)

object ClassNextLessons {
    fun next(week: ScheduleWeek, classId: String, now: LocalDateTime): ClassNextLesson? =
        map(week, now)[classId.lowercase()]

    /** One entry per class id found on the week, keyed lowercased. */
    fun map(week: ScheduleWeek, now: LocalDateTime): Map<String, ClassNextLesson> {
        val grouped = linkedMapOf<String, MutableList<ScheduleEvent>>()
        for (day in week.days) {
            for (event in day.events) {
                val classId = ClassRoster.classId(event.href, event.team) ?: continue
                val start = event.start ?: continue
                if (event.isAllDay) continue
                if (event.status == EventStatus.CANCELLED) continue
                grouped.getOrPut(classId.lowercase()) { mutableListOf() }.add(event)
            }
        }
        val result = linkedMapOf<String, ClassNextLesson>()
        for ((id, events) in grouped) {
            val picked = pick(events, now) ?: continue
            result[id] = ClassNextLesson(
                start = picked.start ?: now,
                room = picked.room?.trim()?.takeIf { it.isNotEmpty() },
            )
        }
        return result
    }

    private fun pick(events: List<ScheduleEvent>, now: LocalDateTime): ScheduleEvent? {
        val ordered = events.sortedBy { it.start }
        return ordered.firstOrNull { event ->
            val start = event.start ?: return@firstOrNull false
            !start.isBefore(now)
        } ?: ordered.firstOrNull()
    }
}

File: ios/BetterW4/ClassNextLesson.swift
//
//  ClassNextLesson.swift
//  BetterW4
//
//  The next (or this week's first) timetable block for a class, from a week
//  that is already in the cache. No network.
//

import Foundation

struct ClassNextLesson: Equatable, Sendable {
    let start: Date
    let room: String?

    /// `Today 10:00` when the block is today in Oslo, otherwise `Wed 10:00`.
    func dayTimeLabel(now: Date) -> String {
        let time = W4Dates.formatTime(start)
        if W4Dates.isSameDay(start, now) {
            return "Today \(time)"
        }
        let weekday = W4Dates.weekdayName(of: start)
        let short = weekday.count >= 3 ? String(weekday.prefix(3)) : weekday
        return "\(short) \(time)"
    }

    func detailLabel(now: Date) -> String {
        let dayTime = dayTimeLabel(now: now)
        if let room, !room.isEmpty {
            return "\(dayTime) · \(room)"
        }
        return dayTime
    }
}

enum ClassNextLessons {
    /// Earliest block for `classId` with `start >= now`, else the first block
    /// of the week. `nil` when this week has no linked class brick.
    static func next(in week: ScheduleWeek, classId: String, now: Date) -> ClassNextLesson? {
        map(in: week, now: now)[classId.lowercased()]
    }

    /// One entry per class id found on the week, keyed lowercased.
    static func map(in week: ScheduleWeek, now: Date) -> [String: ClassNextLesson] {
        var grouped: [String: [TimetableEvent]] = [:]
        for event in week.allEvents {
            guard let classId = ClassRoster.classId(from: event.href) else { continue }
            guard let start = event.start, !event.isAllDay else { continue }
            if event.status == .cancelled { continue }
            grouped[classId.lowercased(), default: []].append(event)
        }
        var result: [String: ClassNextLesson] = [:]
        for (id, events) in grouped {
            guard let picked = pick(events, now: now) else { continue }
            result[id] = ClassNextLesson(
                start: picked.start ?? now,
                room: picked.room?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
            )
        }
        return result
    }

    private static func pick(_ events: [TimetableEvent], now: Date) -> TimetableEvent? {
        let ordered = events.sorted { lhs, rhs in
            (lhs.start ?? .distantFuture) < (rhs.start ?? .distantFuture)
        }
        if let upcoming = ordered.first(where: { ($0.start ?? .distantPast) >= now }) {
            return upcoming
        }
        return ordered.first
    }
}
